erpnext/accounts/utils.py

```diff
@@ -1508,24 +1508,29 @@ def get_stock_accounts(company, voucher_type=None, voucher_no=None):
 				)
 			]
 
-	return stock_accounts
+	return list(set(stock_accounts))
 
 
 def get_stock_and_account_balance(account=None, posting_date=None, company=None):
 	if not posting_date:
 		posting_date = nowdate()
 
-	warehouse_account = get_warehouse_account_map(company)
-
 	account_balance = get_balance_on(
 		account, posting_date, in_account_currency=False, ignore_account_permission=True
 	)
 
-	related_warehouses = [
-		wh
-		for wh, wh_details in warehouse_account.items()
-		if wh_details.account == account and not wh_details.is_group
-	]
+	total_stock_accounts = frappe.get_all(
+		"Account", {"account_type": "Stock", "company": company, "is_group": 0}
+	)
```

Comment:
can this be done like so?

```py
set(warehouse.account for warehouse in get_warehouse_account_map(company).values())
```

Reply:
No need to execute this method if the single stock account exists because in that case need to consider all warehouses. So better if the single stock account is then, remove warehouse condition. 
Also we don't have to get the warehouse account but we want warehouses from the get_warehouse_account_map method

Reply:
The snippet above is basically to identify the unique number of stock accounts linked to warehouses. The difference between my snippet above and `get_all` is that my snippet uses `get_warehouse_account_map` which is cached.

Reply:
get_warehouse_account_map will be slow if there will be single account multiple warehouses. The person who reported this issue has more than 10,800 warehouses but single Stock Account.

Maybe instead of using frappe.get_all, I will run simply count query.

Reply:
Not something you need to look in for this PR, but `get_warehouse_account_map` can also be easily simplified. If we get `lft` & `rgt` in the query, we can easily set the accounts as well in child warehouses. The main issue is queries inside loops.

Reply:
Yes, that I will fix